sk-gen/src/main.rs

```diff
@@ -0,0 +1,700 @@
+        // petgraph may have internal capabilities for this, but I haven't had the time to look
+        // if this stays a part of our code, we may want to wrap the graph w/ tracking data in a new struct
+        // -HM
+        let mut node_to_index: HashMap<Node, NodeIndex> = HashMap::new();
+        node_to_index.insert(starting_state.clone(), starting_node_idx);
+
+        // To find the graph containing all valid traces of trace_length with a given start state, we
+        // perform bfs to a depth of trace_length. Queue item: `(depth, deployment)`
+        let mut bfs_queue: VecDeque<(u64, Node)> = VecDeque::new();
+        bfs_queue.push_back((1, starting_state)); // start at depth 1
+        let mut visited = HashSet::new();
+
+        while let Some((depth, node)) = bfs_queue.pop_front() {
+            let node_idx = *node_to_index.get(&node).expect("node not found in node_to_index");
+
+            if depth >= trace_length {
+                continue;
+            }
+
+            let not_previously_seen = visited.insert(node.clone());
+            assert!(not_previously_seen);
+
+            node.valid_action_states(&cluster_graph.candidate_deployments)
+                .into_iter()
+                .for_each(|(action, next_state)| {
+                    let next_idx = *node_to_index.entry(next_state.clone()).or_insert_with(|| {
+                        let node = cluster_graph.graph.add_node(next_state.clone());
+                        bfs_queue.push_back((depth + 1, next_state.clone()));
+                        node
+                    });
+
+                    // We precompute the trace_event once here for our edge rather than recomputing it every
+                    // time the edge is traversed in a walk.
+                    let trace_event = gen_trace_event(BASE_TS + depth as i64, &node, &next_state);
+
+                    // Because we are not revisiting outgoing nodes, we can be sure that the edge does not already exist
+                    // so long as the same (node, node) edge is not achievable by distinct actions
+                    cluster_graph
+                        .graph
+                        .update_edge(node_idx, next_idx, Edge { action, trace_event });
+                });
+        }
+
+        cluster_graph
+    }
+
+    /// Generate all walks of `trace_length` starting from the first node in the graph.
+    ///
+    /// Returns a list of [`Walk`]s, where each is a list of `(incoming edge, node)` pairs.
+    /// The first node of each walk, and thus the first pair, has no incoming edge, but all
+    /// remaining pairs contain `Some` edge.
+    fn generate_walks(&self, trace_length: u64) -> Vec<Walk> {
+        let walk_start_node = self.graph.node_indices().next().unwrap();
+
+        // We use a depth-first search because eventually we may want to use stochastic methods which do not
+        // fully enumerate the neighborhood of each visited node.
```

Comment:
Is this implying that eventually we want to no longer produce complete graphs but only generate certain neighbors of a surrounding region?

Reply:
Yes! Potentially. This goes along with your other comment of not generating all the actions.